docs/End-To-End-Encryption.md

```diff
@@ -0,0 +1,181 @@
+# End-to-End Encryption
+
+## Standard ECIES Based End-to-End Encryption
+
+PowerAuth supports a standard ECIES encryption (integrated encryption scheme that uses elliptic curve cryptography) with the standard X9.63 (SHA256) KDF function (that produces 32b long keys).
+
+### ECIES Encryption
+
+Assume we have a public key `KEY_ENC_PUB`, data `DATA_ORIG` to be encrypted and a `SHARED_INFO_1` and `SHARED_INFO_2` constants (`byte[]`) as an encryption parameters. ECIES encryption works in a following way:
+
+1. Generate an ephemeral key pair:
+    ```java
+    EPH_KEYPAIR = (KEY_EPH_PRIV, KEY_EPH_PUB).
+    ```
+2. Derive base secret key (in this step, we do not trim the key to 16b only, we keep all 32b).
+    ```java
+    SecretKey KEY_BASE = ECDH.phase(KEY_EPH_PRIV, KEY_ENC_PUB)
+    ```
+3. Derive a secret key using X9.63 KDF function (using SHA256 internally). When calling the KDF, we use `SHARED_INFO_1` together with `KEY_EPH_PUB` value (as raw `byte[]`) as an `info` parameter.
+    ```java
+    byte[] INFO = Bytes.concat(SHARED_INFO_1, KEY_EPH_PUB);
+    SecretKey KEY_SECRET = KDF_X9_63_SHA256.derive(KEY_BASE, INFO)
+    ```
+4. Split the 32b long `KEY_SECRET` to two 16b keys. The first part is used as an encryption key `KEY_ENC`. The second part is used as MAC key `KEY_MAC`.
+    ```java
+    SecretKey KEY_ENC = ByteUtils.subarray(KEY_SECRET, 0, 16);
```

Comment:
Add  `keyConverter.convertBytesToSharedSecretKey()`

Reply:
We don't have such primitive defined. We don't go to technical details in "crypto" documentation (except in technical notes). The algorithm is written in kind of pseudo code and should not by complicated by a specific implementation. For example, on Mobile SDK side, there's no such conversion function available, because the key is always an array of bytes.

Reply:
I didn't make it up - we use such method in documentation in various documents:
https://github.com/wultra/powerauth-crypto/pull/235/files#diff-560395e8c5ec5b9e9a2c4ab7c803decdR47
https://github.com/wultra/powerauth-crypto/pull/235/files#diff-2f630db59881beaa3445c63022648d8eR41
https://github.com/wultra/powerauth-crypto/pull/235/files#diff-163b8bff84f4c49fa208704d43ba7178R66
https://github.com/wultra/powerauth-crypto/pull/235/files#diff-faee70cb6e6858c95feda286156d8638R25

I guess it is ok to leave it out, just a bit inconsistent, because we do the conversion in all other cases.

Reply:
OK, I see the problem. We're using `SecretKey` (almost everywhere) and `SharedKey` (just few definitions). First one has a conversion routine defined, 2nd one not. I'll replace `SharedKey` with `SecretKey` and use that conversion routine in this situation.